File: fabric/src/main/java/com/vandendaelen/nicephore/fabric/mixin/ScreenshotMixin.java
package com.vandendaelen.nicephore.fabric.mixin;

import com.mojang.blaze3d.platform.NativeImage;
import com.vandendaelen.nicephore.fabric.platform.FabricScreenshotHook;
import kotlin.jvm.functions.Function2;
import net.minecraft.client.Screenshot;
import net.minecraft.network.chat.Component;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

import java.io.File;
import java.util.function.Consumer;

/**
 * Intercepts screenshot capture on Fabric. fabric-api 0.145.4+26.1.2 does not expose a
 * ScreenshotEvent API in the version we ship against, so we inject at the head of
 * Screenshot.lambda$grab$1 — the Runnable submitted to the IO pool that writes the file
 * and posts the result Component. Target is a static synthetic method, so the handler
 * must also be static (written in Java for predictable Mixin bytecode behaviour).
 *
 * <p><b>Fragility:</b> the target method name {@code lambda$grab$1} is compiler-generated
 * and may change in future Minecraft versions (renumbering, refactor). Mitigations:
 * <ul>
 *   <li>{@code required: false} and {@code defaultRequire: 0} in the mixin config so a
 *       failed apply is a log warning, not a fatal boot error.</li>
 *   <li>A reflection probe in {@link FabricScreenshotHook#probeTargetMethodPresent()}
 *       checks at boot whether the target still exists and logs explicitly if it doesn't.</li>
 *   <li>The screenshot capture silently becomes a no-op if the mixin fails to apply; users
 *       still get vanilla screenshots (just without Nicephore's custom chat message, clipboard
 *       copy, and optimisation pipeline on Fabric).</li>
 * </ul>
 */
@Mixin(Screenshot.class)
public abstract class ScreenshotMixin {

    @Inject(
            method = "lambda$grab$1(Lcom/mojang/blaze3d/platform/NativeImage;Ljava/io/File;Ljava/util/function/Consumer;)V",
            at = @At("HEAD"),
            cancellable = true,
            remap = false,
            require = 0
    )
    private static void nicephore$onGrabLambda1(
            NativeImage image,
            File file,
            Consumer<Component> msgConsumer,
            CallbackInfo ci
    ) {
        Function2<NativeImage, File, Component> callback = FabricScreenshotHook.getRegistered();
        if (callback == null) {
            return;
        }
        try {
            image.writeToFile(file);
        } catch (Throwable t) {
            return;
        }
        Component replacement = callback.invoke(image, file);
        if (replacement != null && !replacement.getString().isEmpty()) {
            msgConsumer.accept(replacement);
        }
        ci.cancel();
    }
}

File: fabric/src/main/kotlin/com/vandendaelen/nicephore/fabric/mixin/ScreenshotMixin.kt
package com.vandendaelen.nicephore.fabric.mixin

import com.mojang.blaze3d.platform.NativeImage
import com.vandendaelen.nicephore.fabric.platform.FabricScreenshotHook
import net.minecraft.client.Screenshot
import net.minecraft.network.chat.Component
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import java.io.File
import java.util.function.Consumer

/**
 * Intercepts screenshot capture on Fabric (no screenshot event in fabric-api 0.145.4+26.1.2).
 *
 * Vanilla flow inside Screenshot.grab:
 *  grab(File, String, RenderTarget, int, Consumer<Component>)
 *   -> takeScreenshot(target, width, Consumer<NativeImage>)
 *      NativeImage consumer = lambda$grab$0(workDir, fileName, msgConsumer, image)
 *   -> lambda$grab$0 builds the output File, then submits Runnable to TracingExecutor:
 *        lambda$grab$1(NativeImage image, File file, Consumer<Component> msgConsumer)
 *   -> lambda$grab$1: image.writeToFile(file); builds success Component; msgConsumer.accept(comp)
 *
 * Strategy: inject at HEAD of lambda$grab$1 (cancellable).
 * If the hook is registered we take over completely:
 *   1. image.writeToFile(file) — save the screenshot
 *   2. callback(image, file) — notify the hook (launches ScreenshotThread, etc.)
 *   3. If callback returns a non-null, non-empty Component → deliver it via msgConsumer
 *      If callback returns Component.literal("") → suppress the chat message
 *   4. cancel() — prevents vanilla from re-saving and re-messaging
 *
 * If the hook is NOT registered, we do nothing and vanilla proceeds normally.
 */
@Mixin(Screenshot::class)
abstract class ScreenshotMixin {

    @Inject(
        method = ["lambda\$grab\$1(Lcom/mojang/blaze3d/platform/NativeImage;Ljava/io/File;Ljava/util/function/Consumer;)V"],
        at = [At("HEAD")],
        cancellable = true,
        remap = false,
    )
    private fun onGrabLambda1(
        image: NativeImage,
        file: File,
        @Suppress("UNCHECKED_CAST") msgConsumer: Consumer<Component>,
        ci: CallbackInfo,
    ) {
        val callback = FabricScreenshotHook.registered ?: return
        // Save the image ourselves
        runCatching { image.writeToFile(file) }.onFailure { return }
        // Notify the hook; it may return a replacement Component (or null to keep vanilla msg)
        val replacement = callback(image, file)
        if (replacement != null && replacement.string.isNotEmpty()) {
            msgConsumer.accept(replacement)
        }
        // Cancel vanilla body whether or not there is a replacement component, because
        // we already wrote the file and the hook has been notified. Vanilla would
        // re-write the file (double-write) and deliver a redundant chat message.
        ci.cancel()
    }
}
